src/state/slices/actionSlice/actionSlice.ts

```diff
@@ -0,0 +1,90 @@
+import type { PayloadAction } from '@reduxjs/toolkit'
+import { createSlice } from '@reduxjs/toolkit'
+import { uuidv4 } from '@walletconnect/utils'
+
+import type { Action, ActionCenterState, ActionPayload, ActionUpdatePayload } from './types'
+import { isLimitOrderPayloadDiscriminator, isTradePayloadDiscriminator } from './types'
+
+export const initialState: ActionCenterState = {
+  actions: [],
+}
+
+export const actionCenterSlice = createSlice({
+  name: 'actionCenter',
+  initialState,
+  reducers: create => ({
+    upsertAction: create.reducer((state, { payload }: PayloadAction<ActionPayload>) => {
+      // as an extra safety avoid deduplicating actions with the same swapId as they should be unique
+      if (isTradePayloadDiscriminator(payload) && payload.metadata?.swapId) {
+        const existingActionIndex = state.actions.findIndex(
+          action =>
+            isTradePayloadDiscriminator(action) &&
+            action.metadata.swapId === payload.metadata.swapId,
+        )
+
+        if (existingActionIndex !== -1) {
+          return
+        }
+      }
+
+      const actionWithId: Action = {
+        ...payload,
+        id: uuidv4(),
+        createdAt: Date.now(),
+        updatedAt: Date.now(),
+      }
+
+      state.actions.push(actionWithId)
+    }),
+    updateAction: create.reducer((state, { payload }: PayloadAction<ActionUpdatePayload>) => {
+      if (
+        !payload.id &&
+        !isTradePayloadDiscriminator(payload) &&
+        !isLimitOrderPayloadDiscriminator(payload)
+      )
+        return
+
+      const actionIndex = (() => {
+        if (isLimitOrderPayloadDiscriminator(payload)) {
+          return state.actions.findIndex(
+            action =>
+              isLimitOrderPayloadDiscriminator(action) &&
+              ((action.metadata.limitOrderId &&
+                action.metadata.limitOrderId === payload.metadata.limitOrderId) ||
+                (action.metadata.quoteId && action.metadata.quoteId === payload.metadata.quoteId)),
+          )
+        }
+
+        if (isTradePayloadDiscriminator(payload)) {
+          return state.actions.findIndex(
+            action =>
+              isTradePayloadDiscriminator(action) &&
+              action.metadata.swapId === payload.metadata.swapId,
+          )
+        }
+
+        return -1
+      })()
+
+      if (actionIndex === -1) return
+
+      const existingAction = state.actions[actionIndex]
+
+      const action = {
+        ...existingAction,
+        ...payload,
+        metadata: {
+          ...existingAction.metadata,
+          ...payload.metadata,
+        },
+        updatedAt: Date.now(),
+        // ts is borked there, couldn't find any cleaner solution than a cast for simplicity
```

Comment:
if TS is drunk here it probably means we shouldn't do this 
<img width="1477" alt="image" src="https://github.com/user-attachments/assets/74d39da4-04ba-4cf9-ab14-38bdb6a9e43c" />

If we have strongly typed action-specific metadata, or a bunch of optional swapper-specific properties like `swapperSpecificMetadata` then this should be automagically solved